hedera-mirror-rest-java/src/main/java/com/hedera/mirror/restjava/service/Bound.java

```diff
@@ -0,0 +1,120 @@
+import com.hedera.mirror.restjava.common.EntityIdRangeParameter;
+import com.hedera.mirror.restjava.common.RangeOperator;
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.List;
+import lombok.Getter;
+import org.springframework.util.CollectionUtils;
+
+public class Bound {
+
+    @Getter
+    private EntityIdRangeParameter lower;
+
+    @Getter
+    private EntityIdRangeParameter upper;
+
+    private final EnumMap<RangeOperator, Integer> cardinality = new EnumMap<>(RangeOperator.class);
+
+    public Bound(List<EntityIdRangeParameter> params, boolean primarySortField) {
+
+        if (CollectionUtils.isEmpty(params)) {
+            return;
+        }
+        for (EntityIdRangeParameter param : params) {
+
+            if (param.hasLowerBound()) {
+                lower = param;
+            } else if (param.hasUpperBound()) {
+                upper = param;
+            }
+            cardinality.merge(param.operator(), 1, Math::addExact);
+        }
+        long adjustedLower = adjustLowerBound();
+        long adjustedUpper = adjustUpperBound();
+
+        if (primarySortField && adjustedLower > adjustedUpper) {
+            throw new IllegalArgumentException("Invalid range provided");
+        }
+    }
+
+    private long adjustUpperBound() {
+        if (this.upper == null) {
+            return Long.MAX_VALUE;
+        }
+        long upperBound = this.upper.value().getId();
+        if (this.upper.operator() == RangeOperator.LT) {
+            upperBound--;
+        }
+        return upperBound;
+    }
+
+    private long adjustLowerBound() {
+        if (this.lower == null) {
+            return 0;
+        }
+        long lowerBound = this.lower.value().getId();
+        if (this.lower.operator() == RangeOperator.GT) {
+            lowerBound++;
+        }
+        return lowerBound;
+    }
+
+    public int getCardinality(RangeOperator... operators) {
+        return Arrays.stream(operators)
+                .mapToInt(x -> cardinality.getOrDefault(x, 0))
+                .sum();
+    }
+
+    public boolean isEmpty() {
+        return lower == null && upper == null;
+    }
+
+    public boolean hasLowerAndUpper() {
+        return lower != null && upper != null;
+    }
+
+    public boolean hasEqualBounds() {
+        return getLower().value().getId() == getUpper().value().getId();
+    }
+
+    public void verifyUnsupported(RangeOperator unsupportedOperator) {
+        if (getCardinality(unsupportedOperator) > 0) {
+            throw new IllegalArgumentException(String.format("Unsupported range operator %s", unsupportedOperator));
+        }
+    }
+
+    public void verifySingleOccurrence() {
+        if (this.getCardinality(RangeOperator.GT, RangeOperator.GTE) > 1
+                || this.getCardinality(RangeOperator.LT, RangeOperator.LTE) > 1
+                || this.getCardinality(RangeOperator.EQ) > 1) {
+            throw new IllegalArgumentException("Single occurrence only supported.");
+        }
+    }
```

Comment:
This is an ambiguous error presented to the user. We can refactor it to make it clearer which operator is at fault. We may also consider passing in the query parameter name into the constructor to provide more accurate error messages.

```suggestion
        verifySingleOccurrence(RangeOperator.EQ);
        verifySingleOccurrence(RangeOperator.GT, RangeOperator.GTE);
        verifySingleOccurrence(RangeOperator.LT, RangeOperator.LTE);
    }
    
    private void verifySingleOccurrence(RangeOperator... rangeOperators) {
        if (this.getCardinality(rangeOperators) > 1) {
            throw new IllegalArgumentException("Only one range operator from %s is allowed for the given parameter".formatted(Arrays.toString(rangeOperators));
        }
    }
```